src/kbmod/search/raw_image_eigen.cpp

```diff
@@ -0,0 +1,767 @@
+  extern "C" void deviceConvolve(float* source_img, float* result_img,
+                                 int width, int height, float* psf_kernel,
+                                 int psf_size, int psf_dim, int psf_radius,
+                                 float psf_sum);
+#endif
+
+
+  void RawImageEigen::convolve(PSF psf) {
+#ifdef HAVE_CUDA
+    deviceConvolve(image.data(), image.data(),
+                   get_width(), get_height(), psf.data(),
+                   psf.get_size(), psf.get_dim(), psf.get_radius(), psf.get_sum());
+#else
+    convolve_cpu(psf);
+#endif
+  }
+
+
+  void RawImageEigen::apply_bitmask(const ImageI& bitmask) {
+    image = (bitmask.array() == 1).select(NO_DATA, image);
+  }
+
+
+  void RawImageEigen::apply_mask1(const ImageI& mask, const int flags) {
+    // I'm not sure a copy is avoidable here, we can pass a non-const Ref
+    // and then edit the mask in-place, but that would edit the Python-side
+    // numpy array as well....
+    auto bitmask = mask.unaryExpr( [&flags](int x){return flags & x;} );
+    apply_bitmask(bitmask);
+  }
+
+
+  // Imma be honest, this function makes no sense to me... I think I transcribed
+  // it right, but I'm not sure? If a flag is in mask, it's masked unless it's in
+  // exceptions? But why do we have two lists? Is the example above better than this?
+  void RawImageEigen::apply_mask(int flags, const std::vector<int>& exceptions,
+                                 const RawImageEigen& mask) {
+    const int npixels = get_npixels();
+    assert(npixels == mask.get_npixels());
+
+    // evaluated as lvalue because of auto (hopefully)
+    auto mask_unravelled = mask.image.reshaped();
+    auto image_unravelled = image.reshaped();
+
+    for (unsigned int p = 0; p < npixels; ++p) {
+      int pix_flags = static_cast<int>(mask_unravelled(p));
+      bool is_exception = false;
+      for (auto& e : exceptions) is_exception = is_exception || e == pix_flags;
+      if (!is_exception && ((flags & pix_flags) != 0)) image_unravelled(p) = NO_DATA;
+    }
+  }
+
+
+  /* This implementation of grow_mask is optimized for steps > 1
+     (which is how the code is generally used. If you are only
+     growing the mask by 1, the extra copy will be a little slower.
+  */
+  void RawImageEigen::grow_mask(const unsigned steps) {
+    const int num_array = width * height;
+
+    // create a bitmask out of the image array
+    ImageI bitmask = ImageI::Zero(height, width);
+    bitmask = (image.array() == NO_DATA).select(1, bitmask);
+
+    // convolution with a dilating operator is probably the
+    // optimal solution but the iterative process of the
+    // previous implementation makes me think it wouldn't
+    // yield the same solution, because the image edits
+    // itself each iteration
+    for (unsigned x=0; x<width; ++x){
+      for (unsigned y=0; y<height; ++y){
+        Index idx(x, y);
+        auto [i, j, p, q] = idx.centered_block(steps, width, height);
```

Comment:
The way grow mask was originally defined was not as a 2d block, but rather a Manhattan distance of `steps`, so if `steps` = 2 and pixel (1, 1) was masked you'd start with a binary mask of:

```
0 0 0 0 0 
0 0 1 0 0
0 0 0 0 0
0 0 0 0 0
0 0 0 0 0
```

and grow the mask to:

```
0 1 1 1 0 
1 1 1 1 1
0 1 1 1 0
0 0 1 0 0
0 0 0 0 0
```

Reply:
Reverted to old behavior. Same as above. In some places I certainly didn't believe it would make a measurable difference to the results, if we provided a more conventional way of doing things. This is how masking was done in Python in ImageCollection for example - when there were no complaints.